kernel/src/history_manager/search.rs

```diff
@@ -0,0 +1,253 @@
+///
+/// 1. For [`Bound::LeastUpper`], find the smallest index `i` such that `values[i] >= key`
+/// 2. For [`Bound::GreatestLower`], find the largest index `i` such that `values[i] <= key`
+///
+/// # Arguments
+///
+/// * values - The slice to search in
+/// * key - The key value to search for
+/// * key_fn - A fallible function that extracts a comparable key from slice elements
+/// * bound - The type of bound to search for (least upper bound or greatest lower bound)
+///
+/// # Returns
+///
+/// * Ok(usize) - The index of the found element
+/// * [`SearchError::OutOfRange`] - If no suitable bound exists in the slice
+/// * [`SearchError::KeyFunctionError`] - If the key function fails. This contains the error
+///   emitted by the `key_fn`.
+///
+/// # Examples
+///
+/// ## Finding the Least Upper Bound
+/// ```rust
+/// # use delta_kernel::history_manager::search::*;
+/// # use delta_kernel::DeltaResult;
+/// let values = [10, 20, 30, 40, 50];
+/// // Simple key function that just returns the value
+/// let key_fn = |&val| -> DeltaResult<_> { Ok(val) };
+///
+/// // Find the least upper bound for 25 (element that is ≥ 25)
+/// let result = binary_search_by_key_with_bounds(
+///     &values,
+///     25,
+///     key_fn,
+///     Bound::LeastUpper
+/// );
+/// assert_eq!(result.unwrap(), 2); // Index of 30
+/// ```
+///
+/// ## Finding the Greatest Lower Bound
+/// ```rust
+/// # use delta_kernel::history_manager::search::*;
+/// # use delta_kernel::DeltaResult;
+/// let values = [10, 20, 30, 40, 50];
+/// let key_fn = |&val| -> DeltaResult<_> { Ok(val) };
+///
+/// // Find the greatest lower bound for 25 (element that is ≤ 25)
+/// let result = binary_search_by_key_with_bounds(
+///     &values,
+///     25,
+///     key_fn,
+///     Bound::GreatestLower
+/// );
+/// assert_eq!(result.unwrap(), 1); // Index of 20
+/// ```
+///
+/// ## Handling Out of Range Values
+/// ```rust
+/// # use delta_kernel::history_manager::search::*;
+/// # use delta_kernel::DeltaResult;
+/// let values = [10, 20, 30, 40, 50];
+/// let key_fn = |&val| -> DeltaResult<_> { Ok(val) };
+///
+/// // Finding a bound that's out of range
+/// let result = binary_search_by_key_with_bounds(
+///     &values,
+///     5,
+///     key_fn,
+///     Bound::GreatestLower
+/// );
+/// assert!(matches!(result, Err(SearchError::OutOfRange)));
+/// ```
+///
+/// ## Using a Fallible Key Function
+/// ```rust
+/// # use delta_kernel::history_manager::search::*;
+/// # use delta_kernel::DeltaResult;
+/// // Using a fallible key function
+/// let values = ["10", "20", "thirty", "40"];
+/// let result = binary_search_by_key_with_bounds(
+///     &values,
+///     25,
+///     |&s| s.parse::<i32>(),
+///     Bound::LeastUpper
+/// );
+/// assert!(matches!(result, Err(SearchError::KeyFunctionError(_))));
+/// ```
+#[internal_api]
+#[allow(unused)]
+pub(crate) fn binary_search_by_key_with_bounds<'a, T, K: Ord + Debug, E: Error>(
```

Comment:
mega nit but maybe type param named V makes sense for values?

Reply:
hmm given that the key we're searching by has this lifetime, it seems to make sense to keep it as K? 
`key: K`